### ISSUE TITLE
Using `--mutex network` on Node 18 will cause the waiting process to exit silently after 90s

### ISSUE THREAD

**Issue by bradzacher** [yarnpkg/yarn#8960]
https://nodejs.org/en/blog/release/v18.0.0#http-timeouts

Node 18 added a new behaviour to the `http` server:

> server.headersTimeout which limits the amount of time the parser will wait to receive the complete HTTP headers is now set to 60000 (60 seconds) by default.

Alongside a (not explicitly documented in the release notes) feature - `server.connectionsCheckingInterval` which defaults to 30000 (30 seconds).

It appears that `connectionsCheckingInterval` gets checked before `headersTimeout` - so after `connectionsCheckingInterval + headersTimeout = 90s` the server will consider a connection timed out if it hasn't sent any headers.

The latest v1 of yarn (by design) opens a connection to the mutex-owning process, and sends no data:

https://github.com/yarnpkg/yarn/blob/158d96dce95313d9a00218302631cd263877d164/src/cli/index.js#L450-L458

So on Node18 after 90s the mutex-owning `http` server declares the connection as timed out, and closes it. However this timeout does not trigger the `close` or `error` events - it just triggers [the `end` event](https://nodejs.org/docs/latest-v18.x/api/net.html#event-end)

> Emitted when the other end of the socket signals the end of transmission, thus ending the readable side of the socket

So the socket cleans itself up, then the enclosing promise is no longer reachable, and node GC's it - meaning there are no open handles remaining. This causes the mutex waiting process to just quietly exit with exit code `0`.

TL;DR - if all of these are true, you will get broken installs:
- you use NodeJS v18+
- you use `--mutex network`
- your install(s) takes longer than 90s

The fix is either to specify `headersTimeout: Number.MAX_SAFE_INTEGER` to cause the server to never classify a connection as timed out, or to add a `end` even handler to the socket so that it can restart if it was ended prematurely.

**Comment by tavin** [comment]
Good find @bradzacher, we chased broken builds for many days because of this.

Seems like it would be a trivial PR but.... I guess we just have to patch yarn locally?

**Comment by bradzacher** [comment]
You can always vendor yarnv1 with a local patch fixing it yeah. I.e. add this to your `.yarnrc`:
```
yarn-path ".yarn/patched-yarn.js"
```
When someone runs `yarn install` in your repo yarn will use the local patched version.

At my company we did this until we were able to migrate to a modern package manager (we switched to pnpm because it integrates nicely with our bazel build infra).
